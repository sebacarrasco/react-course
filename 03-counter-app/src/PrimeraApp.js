import React, { Fragment } from 'react';
import PropTypes from "prop-types"; //-> para validaciones de las props

// Estos son Functional Components
// Las props se pueden recibir como todo un objeto
// o desestructurizando -> en ese caso se pueden poner valores por defecto
const PrimeraApp = ( { saludo, parrafo = 'Cómo estamos?' } ) => {

    // Lo de fragment es necesario ya que se TIENE que retornar solo un elemento
    // (que puede contener a otros elementos)
    // Se podría hacer con un div pero estaríamos añadiendo elementos
    // inncesarios. Con el Fragment se ponen "sueltos"
    // return (
    //     <Fragment>
    //         <h1>Holaaa</h1>
    //         <p>Cómo estamos?</p>
    //     </Fragment>
    // );

    const objeto = {mensjaje: "hola", id: 2}

    // Poner un fragment es equivalente a poner etiquetas vacías
    return (
        <>
            <h1>{ saludo } locos</h1>
            <p>{ parrafo }</p>
            <p>Un booleano: {true} </p>
            <p> Un arreglo: { [1, 2, 3, 4] }</p>
            <pre> Un objeto: { JSON.stringify(objeto, null, 3) }</pre>
        </>
    );
    // { true } no mostraría nada en el html (ningún booleano)
    // Poner un {objeto} sin stringify tira error (null es el replacer y 3 es el espaciado -> para que se vea bonito)
}

// Validaciones de las props
PrimeraApp.propTypes = {
    saludo: PropTypes.string.isRequired // Sin el .isRequired, si es que no se pasa un valor, no tira error,
    // Solo tiraría error si es que se pasa un valor y este no es un string
}

// También se pueden poner acá los valores por defecto
// PrimeraApp.defaultProps = {
//     parrafo: "Buena buena"
// }

export default PrimeraApp;