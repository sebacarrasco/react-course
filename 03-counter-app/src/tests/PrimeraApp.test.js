import { render } from "@testing-library/react";
import PrimeraApp from "../PrimeraApp";
import React from 'react';
import { shallow } from "enzyme";

describe('Pruebas en <PrimeraApp /> con las herramientas de testeo que viene por defecto (@testing-library/jest-dom/extend-expect)',
    () => {
    test('Debe mostrar el mensaje Hola hola', () => {
        const saludo = "Hola hola";
        const wrapper = render( <PrimeraApp saludo={ saludo } /> );
        const {getByText} = wrapper; // Recordar que esto es lo mismo que hacer getByText=wrapper.getByText
        // Y las dos líneas anteriores es lo mismo que hacer lo de la siguiente línea:
        // const {getByText} = render(<PrimeraApp saludo={saludo} />);

        expect(getByText(saludo + " locos")).toBeInTheDocument();
    });
    
});


describe('Pruebas en <PrimeraApp /> con las herramientas de testeo de Enzyme', () => {
    test('Debe mostrar <PrimeraApp /> correctamente', () => {

        const saludo = "Hola hola";

        // shallow es como render pero permite otras cosas como simular clicks
        const wrapper = shallow(<PrimeraApp saludo={ saludo } />);

        // Para debuggear: console.log(wrapper.html())

        // La siguiente línea va a fallar siempre la primera vez, ya que el snapshot que está
        // guardado al principio es vacío. Por eso hay que apretar `u` en la consola para
        // updatear el snapshot con el wrapper actual. (esto se puede hacer gracias a enzyme-to -json)
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe mostrar el subtítulo enviado mediante props', () => {
        const saludo = "Hola hola";
        const parrafo ="Wena soy un parrafo";
        const wrapper = shallow(<PrimeraApp saludo={ saludo } parrafo={parrafo}/>);

        // Obtenemos el texto del primer párrafo
        const textoParrafo = wrapper.find('p').first().text(); // selector como jquery

        expect(textoParrafo).toBe(parrafo);
    });
    
    
});