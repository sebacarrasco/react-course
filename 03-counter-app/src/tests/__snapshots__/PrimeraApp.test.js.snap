// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <PrimeraApp /> con las herramientas de testeo de Enzyme Debe mostrar <PrimeraApp /> correctamente 1`] = `
<Fragment>
  <h1>
    Hola hola
     locos
  </h1>
  <p>
    CÃ³mo estamos?
  </p>
  <p>
    Un booleano: 
    <Component />
     
  </p>
  <p>
     Un arreglo: 
    1
    2
    3
    4
  </p>
  <pre>
     Un objeto: 
    {
   "mensjaje": "hola",
   "id": 2
}
  </pre>
</Fragment>
`;
